Convert "local & session storage restoration" SPI into API

RupinMittal opened the pull request:
#### 7d45f9b66472677f35f7cc48e486c392720196f0
<pre>
Convert &quot;local &amp; session storage restoration&quot; SPI into API
<a href="https://bugs.webkit.org/show_bug.cgi?id=290418">https://bugs.webkit.org/show_bug.cgi?id=290418</a>
<a href="https://rdar.apple.com/141842219">rdar://141842219</a>

Reviewed by Sihui Liu.

Based on API review, the following changes are being made as
the SPI is promoted to API:

1. &quot;fetchDataOfTypes&quot; will return both an NSData and NSError.

   The data will be nullptr if there is an error and the error
   will be nullptr if there not.

2. The completion handler of &quot;fetchDataOfTypes&quot; will be called with
   an std::optional&lt;HashMap&gt; instead of just a HashMap.

   This is to differentiate between the cases of there being no data
   to return (the optional will contain an empty HashMap) versus an error
   occurring during retrieval (in which case the optional will have no
   value). For example, if the Network Process crashes during the retrieval,
   the completion handler will be called with the default value of an
   optional (std::nullopt) which will indicate to &quot;fetchDataOfTypes&quot; that
   it should return an error.

2. &quot;restoreData&quot; will return an error instead of a boolean.

   For example, if &quot;restoreData&quot; is called with data that doesn&apos;t follow
   the intended format, or the Network Process crashes during restoration,
   an error will be thrown.

The existing SPI will remain for now until users adopt the API. To reduce
code duplication, the existing SPI has been converted to use the API.
The API tests that previously tested the SPI have been converted to test the API.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchLocalStorage):
(WebKit::NetworkProcess::fetchSessionStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::fetchLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView fetchDataOfTypes:completionHandler:]):
(-[WKWebView restoreData:completionHandler:]):
(-[WKWebView _fetchDataOfTypes:completionHandler:]):
(-[WKWebView _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore restoreData:completionHandler:]):
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
(filterSupportedTypes): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::fetchSessionStorage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(testRestoreLocalStorage):
(testRestoreLocalStorageThirdPartyIFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
(testRestoreSessionStorage):
(testRestoreSessionStorageThirdPartyIFrame):

Canonical link: <a href="https://commits.webkit.org/293571@main">https://commits.webkit.org/293571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/724482828264d0c366f5e3d0bf670c5c902fd7f1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18940 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9189 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104422 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27373 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75563 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32669 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102299 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14610 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89644 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55923 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7625 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49252 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84329 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106780 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26405 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19251 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84526 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26768 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84038 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28702 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6388 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20134 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16152 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26345 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31536 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26166 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27733 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3184,6 +3184,118 @@ - (void)_updateFixedColorExtensionViewFrames
+            switch (dataTypeToEncode) {
+            case WebKit::WebViewDataType::SessionStorage:
+                if (!data->sessionStorage) {
+                    completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```
I doubt this is the right error but am not sure what would be better. This will be called if the Network Process crashes and the completion handler is called with the default arguments (`std::nullopt`). Any suggestions?

cdumez replied:
Please null `nullptr` instead of `NULL` everywhere.

RupinMittal commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3184,6 +3184,118 @@ - (void)_updateFixedColorExtensionViewFrames
+        if (*succeeded)
+            completionHandler(NULL);
+        else
+            completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```
I doubt this is the right error but am not sure what would be better. This will be called if the if the restoration fails due to a NetworkProcess crash. Any suggestions?

RupinMittal commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+            switch (dataTypeToEncode) {
+            case WebKit::WebsiteDataType::LocalStorage:
+                if (!data->localStorage) {
+                    completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```
I doubt this is the right error but am not sure what would be better. This will be called if the Network Process crashes and the completion handler is called with the default arguments (std::nullopt). Any suggestions?

RupinMittal commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+        if (*succeeded)
+            completionHandler(NULL);
+        else
+            completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```
I doubt this is the right error but am not sure what would be better. This will be called if the Network Process crashes and the completion handler is called with the default arguments (std::nullopt). Any suggestions?

RupinMittal reviewed (commented):
Had a few questions.

RupinMittal commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
```diff
@@ -666,6 +666,26 @@ The uniform type identifier kUTTypeWebArchive can be used get the related pasteb
+
+/* @abstract Called when the client wants to fetch WKWebView data.
+   @param dataTypes The option set of WKWebView data types whose data the client wants to fetch.
+   @param completionHandler The completion handler that should be invoked with the retrieved data and possibly an error. The retrieved data will be a serialized blob. If an error occurred, the retrieved data will be nil. An error may occur if a requested data type is not supported or if the data cannot be retrieved for some other reason (such as a crash).
```
How should we test for a user requesting an unsupported data type? Is it even possible? They have to use the NS_OPTIONS above.

jesxilin commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3184,6 +3184,118 @@ - (void)_updateFixedColorExtensionViewFrames
+    std::optional<Vector<WebKit::WebViewDataType>> encodedDataTypes;
+    decoder >> encodedDataTypes;
+    if (!encodedDataTypes) {
+        completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:NULL]);
```
I think for this and others, there is a leak whenever you have this `[NSError alloc]...` and you should do something like `completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:NULL]).get())` instead.

cdumez replied:
@jesxilin is totally right. The Safer CPP bot should be able to notice things like this starting next week (cc @rniwa)

RupinMittal replied:
Oh true! Thank you for the catch @jesxilin! Will fix.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -3166,11 +3166,11 @@ void NetworkProcess::setPersistedDomains(PAL::SessionID sessionID, HashSet<Regis
     CheckedPtr session = networkSession(sessionID);
     if (!session) {
-        completionHandler({ });
+        completionHandler(std::nullopt);
```
I am confused by this returns nullopt...

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -3188,11 +3188,11 @@ void NetworkProcess::restoreLocalStorage(PAL::SessionID sessionID, HashMap<WebCo
     CheckedPtr session = networkSession(sessionID);
     if (!session) {
-        completionHandler({ });
+        completionHandler(HashMap<WebCore::ClientOrigin, HashMap<String, String>>());
```
... but this returns an empty HashMap.

The 2 seems like the same case, just one for local storage and the other for session storage. The lack of consistency here is fishy.

RupinMittal replied:
Yeah was intentional but I'm not 100% sure about this. Since session storage is not stored on disk, it seemed that having no network process just means there's no storage so we can return an empty map. But for local storage since it is stored on disk, having no network process means we don't know if there is storage or not and we can't access it--so we should return an error (returning `std::nulltopt` will cause an error to be thrown).

What do we think about that?

szewai replied:
This should return `std::nullopt` too.
We should return empty map for `SessionStorage` if network process doesn't exist, but in this case network process exists (it's handling this message), and it's unexpected running network process does not have information about requested session.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+
+    auto data = Box<WKWebsiteData>::create();
+
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), dataTypesToEncode = WTFMove(dataTypesToEncode), data] {
```
We do not hide smart pointers behind `auto`

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2874,12 +2874,12 @@ void WebsiteDataStore::setRestrictedOpenerTypeForDomainForTesting(const WebCore:
         networkProcess->fetchLocalStorage(m_sessionID, WTFMove(completionHandler));
     else
-        completionHandler({ });
+        completionHandler(std::nullopt);
```
This should be `protectedNetworkProcess()->fetchLocalStorage(m_sessionID, WTFMove(completionHandler));`, to launch network process to fetch data on disk.

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -16252,12 +16252,12 @@ FloatPoint WebPageProxy::mainFrameScrollPosition() const
         networkProcess->sendWithAsyncReply(Messages::NetworkProcess::FetchSessionStorage(sessionID(), identifier()), WTFMove(completionHandler));
     else
-        completionHandler({ });
+        completionHandler(HashMap<WebCore::ClientOrigin, HashMap<String, String>>());
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>> { }`

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+    std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> localStorage;
+};
+
+- (void)fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(void(^)(NSData *, NSError *))completionHandler
```
Please refactor the code so that `fetchDataOfTypes` /  `_fetchDataOfTypes`, and `restoreData` / `_restoreData` will share the same implementation instead of having duplicate code.  E.g. adding a `fetchDataOfTypesShared` function.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3184,6 +3184,118 @@ - (void)_updateFixedColorExtensionViewFrames
+    std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> sessionStorage;
+};
+
+- (void)fetchDataOfTypes:(WKWebViewDataType)dataTypes completionHandler:(void (^)(NSData *, NSError *))completionHandler
```
Please refactor the code so that `fetchDataOfTypes` / ` _fetchDataOfTypes`, and `restoreData` / `_restoreData` will share the same implementation instead of having duplicate code.

RupinMittal replied:
Fixed and added some error messages too.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3192,6 +3192,124 @@ - (void)_updateFixedColorExtensionViewFrames
+    decoder >> currentWKWebViewDataSerializationVersion;
+    if (!currentWKWebViewDataSerializationVersion) {
+        NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Data is in incorrect format.", };
+        completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:userInfo]).get());
```
Why does this return `NSCoderReadCorruptError` error code? `NSCoder` is not used in encoding / decoding (it uses WebKit's persistent encoder).
Can we use `WKErrorUnknown` with more explicit message like "Version is invalid"?

RupinMittal replied:
I've added error messages for the decoding part.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,117 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+        NSString *unsupportedMessage = [unsupportedPrefix stringByAppendingString:dataType];
+        NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : unsupportedMessage, };
+
+        completionHandler(nullptr, adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSFeatureUnsupportedError userInfo:userInfo]).get());
```
Need to early return after completionHandler is invoked.

szewai replied:
We should use `WKErrorCode` for error with `WKErrorDomain` domain

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3192,6 +3192,124 @@ - (void)_updateFixedColorExtensionViewFrames
+
+            if (!sessionStorage) {
+                NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Data is in incorrect format.", };
+                completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:userInfo]).get());
```
Does not seem right to invoke `completionHandler` after it's moved to `callbackAggregator`.
You might construct `WKWebViewData` from the encoded data before creating `CallbackAggregator` with `CompletionHandler` and invoking `restoreSessionStorage`.

Same issue for the API in `WKWebsiteDataStore`.

RupinMittal replied:
Fixed!

RupinMittal replied:
I fixed it by adding an error variable that gets set and used when the callback aggregator is called. So the completion handler is no longer called after it's moved to the callback aggregator.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -605,6 +605,127 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+        NSString *unsupportedMessage = [unsupportedPrefix stringByAppendingString:dataType];
+        NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : unsupportedMessage, };
+
+        completionHandler(nullptr, adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:userInfo]).get());
```
Missing early return

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5824,109 +5952,25 @@ - (_WKWebProcessState)_webProcessState
+        if (!error)
+            completionHandler(YES);
+        else
+            completionHandler(NO);
```
Can we just do `completionHandler(!error)`?

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3192,6 +3192,134 @@ - (void)_updateFixedColorExtensionViewFrames
+    }
+
+    auto succeeded = Box<bool>::create(true);
+    auto error = Box<RetainPtr<NSError>>::create(nil);
```
Do we actually need two variables? Can we use whether error is null to decide if operation succeeds?

RupinMittal replied:
Good point--fixed!

szewai reviewed (approved):
r=me if bots are happy

RupinMittal committed:
Convert "local & session storage restoration" SPI into API
https://bugs.webkit.org/show_bug.cgi?id=290418
rdar://141842219

Reviewed by Sihui Liu.

Based on API review, the following changes are being made as
the SPI is promoted to API:

1. "fetchDataOfTypes" will return both an NSData and NSError.

   The data will be nullptr if there is an error and the error
   will be nullptr if there not.

2. The completion handler of "fetchDataOfTypes" will be called with
   an std::optional<HashMap> instead of just a HashMap.

   This is to differentiate between the cases of there being no data
   to return (the optional will contain an empty HashMap) versus an error
   occurring during retrieval (in which case the optional will have no
   value). For example, if the Network Process crashes during the retrieval,
   the completion handler will be called with the default value of an
   optional (std::nullopt) which will indicate to "fetchDataOfTypes" that
   it should return an error.

2. "restoreData" will return an error instead of a boolean.

   For example, if "restoreData" is called with data that doesn't follow
   the intended format, or the Network Process crashes during restoration,
   an error will be thrown.

The existing SPI will remain for now until users adopt the API. To reduce
code duplication, the existing SPI has been converted to use the API.
The API tests that previously tested the SPI have been converted to test the API.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchLocalStorage):
(WebKit::NetworkProcess::fetchSessionStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchSessionStorageForWebPage):
(WebKit::NetworkStorageManager::fetchLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView fetchDataOfTypes:completionHandler:]):
(-[WKWebView restoreData:completionHandler:]):
(-[WKWebView _fetchDataOfTypes:completionHandler:]):
(-[WKWebView _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore restoreData:completionHandler:]):
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
(filterSupportedTypes): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::fetchSessionStorage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(testRestoreLocalStorage):
(testRestoreLocalStorageThirdPartyIFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreSessionStorage.mm:
(testRestoreSessionStorage):
(testRestoreSessionStorageThirdPartyIFrame):

Canonical link: https://commits.webkit.org/293571@main
